common/src/main/java/com/microsoft/identity/common/internal/platform/DevicePopManager.java

```diff
@@ -615,6 +625,162 @@ public boolean verify(@NonNull final SigningAlgorithm alg,
+            errCode = NO_SUCH_ALGORITHM;
+            exception = e;
+        } catch (final KeyStoreException e) {
+            errCode = KEYSTORE_NOT_INITIALIZED;
+            exception = e;
+        } catch (final NoSuchPaddingException e) {
+            errCode = NO_SUCH_PADDING;
+            exception = e;
+        } catch (final UnsupportedEncodingException e) {
+            errCode = UNSUPPORTED_ENCODING;
+            exception = e;
+        } catch (final IOException e) {
+            errCode = IO_ERROR;
+            exception = e;
+        }
+
+        final ClientException clientException = new ClientException(
+                errCode,
+                exception.getMessage(),
+                exception
+        );
+
+        Logger.error(
+                TAG + methodName,
+                errCode,
+                exception
+        );
+
+        throw clientException;
+    }
+
+    @Override
+    public String decrypt(@NonNull final Cipher cipher,
+                          @NonNull final String ciphertext) throws ClientException {
+        final String methodName = ":decrypt";
+        final String errCode;
+        final Exception exception;
+
+        try {
+            // Load our key material
+            final KeyStore.PrivateKeyEntry privateKeyEntry = (KeyStore.PrivateKeyEntry)
+                    mKeyStore.getEntry(mKeyAlias, null);
+
+            // Get a reference to our private key (will not be loaded into app process)
+            final PrivateKey privateKey = privateKeyEntry.getPrivateKey();
+
+            // Init our cipher instance, don't use a named provider as there seems to be a mix of
+            // BoringSSL & AndroidOpenSSL
+            // https://issuetracker.google.com/issues/37091211
+            final javax.crypto.Cipher outputCipher = javax.crypto.Cipher.getInstance(cipher.toString());
+            outputCipher.init(javax.crypto.Cipher.DECRYPT_MODE, privateKey);
+
+            // Put our ciphertext into an InputStream
+            final CipherInputStream cipherInputStream = new CipherInputStream(
+                    new ByteArrayInputStream(
+                            Base64.decode(
+                                    ciphertext,
+                                    Base64.DEFAULT
+                            )
+                    ),
+                    outputCipher // Our decryption cipher
+            );
+
+            // Declare a List for dynamic sizing
+            final List<Byte> values = new ArrayList<>();
+
+            // Iterate over bytes, adding them to our List
+            int nextByte;
+            while ((nextByte = cipherInputStream.read()) != -1) {
+                values.add((byte) nextByte);
+            }
+
+            final byte[] bytes = new byte[values.size()];
+
+            for (int ii = 0; ii < bytes.length; ii++) {
+                bytes[ii] = values.get(ii);
+            }
+
+            return new String(bytes, 0, bytes.length, ENCODING_SCHEME);
```

Comment:
This is a lot of copying.  Why not jut put it in a InputStreamReader: https://docs.oracle.com/javase/8/docs/api/java/io/InputStreamReader.html ?

Reply:
Lookin' tidier already
84e685554367f6eaaa4d4afccb70e478a944bc0e